export deserializeBlocks method

gnarf opened the pull request:
### Resolves

Export a helper method from serialization/sb3 to take a serialized `blocks` and produce something a bit more "programmatically parseable"

### Proposed Changes

Takes some inner functionality to the larger `deserialize` method and creates `deserializeBlocks` which does that and exports it for external utilities to be able to depend on.

gnarf committed:
export deserializeBlocks method

gnarf committed:
Make importing with deserialized stuff okay

gnarf committed:
Also export serializeBlocks for round-tripping

kchadha commented on src/serialization/sb3.js:
```diff
@@ -690,6 +690,7 @@ const deserializeInputs = function (inputs, parentId, blocks) {
     for (const inputName in inputs) {
         if (!inputs.hasOwnProperty(inputName)) continue;
         const inputDescArr = inputs[inputName];
+        if (!Array.isArray(inputDescArr)) continue;
```
I think I'm still a little hesitant about this change. 

At the very least, this change (and the similar one below) needs a comment to explain that this line is solely for accommodating deserializing deserialized blocks (which is not the normal use case of this function).

In general, I think it's good for functions like this to be idempotent, but one thing that's weird about making deserialize work on already deserialized things is that there isn't really a similar analogy for `serialize` to work that way.

The closest (albeit somewhat different) thing to this is JSON.parse and JSON.stringify... 
`JSON.stringify(JSON.stringify(x))` and `JSON.stringify(x)` are not the same thing, and `JSON.parse(JSON.parse(x))` doesn't work.

gnarf replied:
basically - there is no way to install targets from json without getting passed through this `deserialize`, I think it makes sense to make `deserialize(alreadyDeserialized)` result in no change (as opposed to weird undefined variables and broken code as it does with the current setup)

kchadha replied:
Do you mind adding comments above both of those lines since they are not required for the normal/common usage flow of deserialize?

gnarf committed:
Add comments for deserialization and add tests

gnarf commented:
@kchadha i added those comments and also added some test cases for serializeBlocks / deserializeBlocks

I noticed while working on the testing that a lot of stuff was using `someObj.hasOwnProperty(prop)` and I replaced some of those with just using the `Object.prototype.hasOwnProperty.call(someObj, prop)` rather than assuming that the objects will be connected to the object prototype.

kchadha reviewed (approved):
LG

gnarf commented:
Thanks for working with me on this one!

This had me thinking that it might be useful to write a `scratch-sb3` and `scratch-sb2` serialize/deserialize package on npm for third parties.
